---
title: "Compare time-course gene expression across species"
editor_options: 
  chunk_output_type: console
---

```{r}
dttm <- Sys.Date()

test <- FALSE
```


---

This website contain analyses by Aman and Billu for an on-going project that compares time-series gene expression data across species.

Majority of the data sets used are from the paper by Laloum and Robinson-Rechavi (add reference here).


> Last updated on `r dttm`.

___

# Load data and glimpse


```{r}
library(dplyr)

# read ortho file ---------------------------------------------------------------

dat <- read.csv(
  here::here(
    "./result/combined_oma_v2.csv"
  ),
  na.strings = c("", " ", "NA")
) |> 
  select(
    - human_name
  ) |> 
  tidyr::separate_longer_delim(
    cflo_ID,
    ", "
  ) |> 
  as_tibble()


# summary 1: different combinations ---------------------------------------

dat |> 
  glimpse()
# mutate_all(as.factor) |> summary()
  mutate(
    across(
      !cflo_ID,
      ~ if_else(
        !is.na(.),
        "present",
        "x"
      )
    )
  ) |> 
  group_by(
    across(
      # everything()
      !cflo_ID
    )
  ) |> 
  tally(
    name = "number of Cflo genes"
  ) |> 
  arrange(
    desc(`number of Cflo genes`)
  ) |> 
  # filter(
  #   n_genes > 100
  # ) |> 
  DT::datatable(
    caption = "Genes present across species."
  )
```

# Genes present in all species

```{r}
# summary 2: present in all -----------------------------------------------

sub <- dat |> 
  select(
    cflo_ID,
    everything()
  ) |> 
  filter(
    !if_any(
      everything(),
      ~ is.na(.)
    )
  )
cat(
  "There are", 
  length(unique(na.omit(sub$cflo_ID))), 
  "unique ant genes that are also present in the other 8 species"
)


```

## Some issues:

Not all of the above Cflo genes are 1:1 orthologs, some likely have paralogs in other species.

```{r}
cat(
  "For how many ant genes, there are â‰¥ 2 genes in at least one more species?"
)
sub |> 
  group_by(
    cflo_ID
  ) |> 
  tally() |> 
  mutate(
    n = if_else(
      n == 1,
      "1:1",
      "1:n possible"
    )
  ) |> 
  group_by(
    n
  ) |> 
  tally() |> 
  rename(
    `Cflo:other species (orthology)` = n,
    `Number of genes` = nn
  ) |> 
  kableExtra::kable(
    caption = "Number of 1:1 orthlogs between Cflo and other species"
  )

```

